src/relay/transforms/annotate_target.cc

```diff
@@ -61,20 +67,27 @@ class AnnotateTargetRewriter : public ExprRewriter {
   std::pair<std::string, Array<Expr>> AnnotateArgs(const Array<Expr>& args,
                                                    const std::string& target = "") {
     std::string ref_target = "";
+    Array<Expr> compiler_begins;
     Array<Expr> compiler_ends;
     for (auto arg : args) {
-      std::string arg_target = "default";
+      std::string arg_target = default_target;
       const CallNode* call = arg.as<CallNode>();
 
       if (call && call->op == CompilerBeginOp()) {
         // Argument is already compiler begin node meaning that this is not the first time
         // running this pass, so we simply remove it and will add a new one later.
         ICHECK_EQ(call->args.size(), 1U);
+        // Do not alter existing annotation if not default
+        if (default_target != call->attrs.as<CompilerAttrs>()->compiler) {
+          compiler_begins.push_back(arg);
+        } else {
+          // Remove default
+          compiler_ends.push_back(call->args[0]);
+        }
```

Comment:
It removes default (but preserve non-default) annotations in case if it is not first invocation of the pass

Reply:
My question was about the algorithm. I don't understand why we "remove default (but preserve non-default) annotations in case if it is not first invocation of the pass". Almost all my comments related to this question.

Meanwhile, PR derives another AnnotateTarget implementation to deal with the case that `include_non_call_ops=False`, so you are more like making a new pass and use `include_non_call_ops` to dispatch the one being used. Combining my previous concern about keeping AnnotateTarget simple and straightforward, I would suggest making it a separate pass. For example, we could name the new pass "PruneSimpleRegions" and use it as following:

1. (MergeComposiste) -> AnnotateTarget -> PruneSimpleRegions -> PartitionGraph
2. (MergeComposiste) -> AnnotateTarget -> MergeCompilerRegions -> PruneSimpleRegions -> PartitionGraph

The first case could achieve your goal, and it could also serve the second case to prune subgraphs (compiler regions) after merging. If we take one step further, we could even let users provide a lambda function to determine if a region is "simple", so that it can be customized for different backends such as TensorRT. The implementation of this pass could also be much simpler and I could even help with that if you prefer.

What do you think?

Reply:
Current contract for AnnotateTarget assumes that for non-call op to be promoted under the target of its arguments it should have all the arguments annotated with the same target. If I am not mistaken, the targets can be added at the run-time, and theoretically the situation is possible when after several subsequent invocations of AnnotateTarget a non-call operation will be annotated as default and will not be promoted even when all its arguments have same target. To prevent this the default annotations always removed while specific targets are preserved. However I am agree that the explanation is a bit far-fetched.

As for the second part of your comment, do I understand correctly that the idea is not to alter the AnnotateTarget and instead do the removal of annotations from non-call ops in a subsequent separate pass? I am not sure I quite understand the definition of "simple" and the concept of PruneSimpleRegions as a whole. Could you elaborate this point a bit more, please?

Reply:
Thanks for the explanation. I think I am now clearer about the first part, but we might need to improve the code comments a lot to make it clear to everyone reading this pass.

For the second part, your understanding is correct. As you mentioned, the definition of "simple" is vague, but it's also more general. Since every target needs to specify their own transform pipeline (e.g., https://github.com/apache/incubator-tvm/blob/main/python/tvm/relay/op/contrib/arm_compute_lib.py#L44), we can every BYOC target can define their own "simple". For example, the definition of "simple" in ACL could be a subgraph with non-call ops; the definition of "simple" in TensorRT could be a subgraph without MAC ops. It means something like

```python
    seq = tvm.transform.Sequential(
        [
            transform.InferType(),
            transform.MergeComposite(arm_compute_lib_pattern_table()),
            transform.AnnotateTarget("arm_compute_lib"),
            transform.PruneSimpleRegions(tvm._ffi.get_global_func("relay.ext.arm_compute_lib.prune")),
            transform.PartitionGraph(),
        ]
    )
```

where `tvm._ffi.get_global_func("relay.ext.arm_compute_lib.prune")` is a packed function in C++ that accepts an `AnnotatedRegion` and outputs a boolean, indicating whether the region should be pruned or not.

Reply:
The proposal is likely to be more flexible. But I am afraid that introducing an externally defined method for doing pruning also means introducing a more elaborate checker for IR as it should be clearly stated what is the correctly pruned graph and what is not.

Reply:
IMHO, it should be fine, because we have abstracted the region to a separate data structure called "AnnotatedRegion". If we pass an AnnotatedRegion to PruneSimpleRegions, we just work on the region level. We can add an API such as "ChangeRegionTarget" to AnnotatedRegion that changes all compiler attributes of the nodes in that region. In this case, PruneSimpleRegion could just do (in Pseudo code):

```
for region in regions:
    if region.target != "default" and is_simple_region(region):
        region.change_region_target("default")
```

and one implementation of "is_simple_region" coud be like:

```
def is_simple_region(region):
    class Checker(ExprVisitor):
        def __init__(self):
            self.is_simple = True
        def visit(self, call):
            self.is_simple = False
            super(self).visit(call)

    checker = Checker()
    for out in region.outs:
        checker.visit(out)
    return checker.is_simple
```

Reply:
I think we need to separate the (valid) concerns about the design and overall maintainability of AnnotateTarget from the intention of this PR. Currently, we have user facing errors with ACL because of this tuple annotation bug and so I think our first priority should be merging something which fixes that. We can discuss possible refactors/redesigns of AnnotateTarget separately with an RFC on the forum where more people can see it and participate.

Reply:
@comaniac Bump

Reply:
Aggregate the discussions to https://github.com/apache/tvm/pull/6655#pullrequestreview-538712328